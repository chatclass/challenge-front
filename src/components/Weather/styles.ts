import styled from 'styled-components'

export const Wrapper = styled.div`
  padding: 1rem 2rem;
  display: flex;
  flex-direction: column;
  position: relative;
  background-color: ${({ theme }) => `${theme.colors.neutrals.white[400]}E6`};
  margin-bottom: 2rem;
  box-shadow: 0px 0px 0.5rem rgba(0, 0, 0, 0.25);

  & > h2 {
    font-size: 2.5rem;
    margin: 1.5rem 0;
    color: ${({ theme }) => theme.colors.neutrals.gray[600]};
  }

  & * svg {
    color: ${({ theme }) => theme.colors.primary.orange};
  }

  @media (min-width: ${({ theme }) => theme.breakpoints.md}) {
    padding: 2rem 4rem;

    & > h2 {
      font-size: 3.5rem;
    }
  }
`

export const Close = styled.button`
  border: 0;
  background: transparent;
  position: absolute;
  padding: 0.25rem;
  top: 0.5rem;
  right: 0.5rem;
  cursor: pointer;
  border: 2px transparent;

  &:focus-visible {
    outline-color: ${({ theme }) => theme.colors.primary.orange};
  }

  @media (min-width: ${({ theme }) => theme.breakpoints.md}) {
    top: 1rem;
    right: 1rem;
  }
`

export const Item = styled.div`
  display: flex;
  gap: 0.5rem;
  align-items: center;

  & > span {
    color: ${({ theme }) => theme.colors.neutrals.gray[500]};
    font-size: 1rem;

    & > b {
      font-weight: 700;
      color: ${({ theme }) => theme.colors.neutrals.black[500]};
    }
  }
`

export const Items = styled.div`
  max-width: 16.25rem;
  display: flex;
  gap: 2rem;
  margin-top: 1rem;
  align-items: center;
  justify-content: space-between;
`

export const Hr = styled.hr`
  width: 100%;
  height: 1px;
  margin: 1rem 0;
  border-top: 0;
  border-bottom: 1px solid ${({ theme }) => theme.colors.primary.orange};

  @media (min-width: ${({ theme }) => theme.breakpoints.md}) {
    margin: 2rem 0;
  }
`

export const Forecast = styled.div`
  display: flex;
  justify-content: space-between;
`

export const Day = styled.div`
  display: flex;
  gap: 0.5rem;
  align-items: center;
  flex-direction: column;

  & > span {
    color: ${({ theme }) => theme.colors.primary.orange};
    font-size: 1rem;
  }

  & > b {
    font-weight: 700;
    color: ${({ theme }) => theme.colors.neutrals.black[500]};
  }
`
