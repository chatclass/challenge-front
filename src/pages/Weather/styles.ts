import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    max-width: ${theme.breakpoints.sm};
    margin: 0 auto;
  `}
`

export const Container = styled.div`
  padding: 0 1.5rem;
  margin: 0 auto;
  justify-content: center;
  display: flex;
  flex-direction: column;
  justify-items: space-between;

  & > h1 {
    font-size: 2.5rem;
    margin: 1.5rem 0;
  }

  & > h3 {
    font-size: 2rem;
  }

  @media (min-width: ${({ theme }) => theme.breakpoints.md}) {
    width: 100%;
    padding: 0;

    & > h1 {
      font-size: 4.5rem;
      margin: 3rem 0;
    }

    & > h3 {
      font-size: 2.5rem;
    }
  }
`

export const Hr = styled.hr`
  width: 100%;
  height: 1px;
  margin: 1rem 0;
  border-top: 0;
  border-bottom: 1px solid white;

  @media (min-width: ${({ theme }) => theme.breakpoints.md}) {
    margin: 2rem 0;
  }
`

export const Dl = styled.dl`
  margin: 2rem 0;
  display: flex;
  flex-direction: column;
  gap: 2rem;
  font-size: large;

  @media (min-width: ${({ theme }) => theme.breakpoints.md}) {
    flex-direction: row;
    gap: 5rem;
  }
`

export const Dt = styled.dt`
  display: flex;
  gap: 2rem;
  font-weight: 300;
`

export const Dd = styled.dd`
  display: flex;
  gap: 2rem;
  margin-top: 1rem;
  font-weight: 700;
`
