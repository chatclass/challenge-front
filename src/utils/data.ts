/* eslint-disable @typescript-eslint/no-explicit-any */
const LIMIT = 6

export const normalizeData = (results: any) => {
  return {
    city: results.city,
    temperature: `${results.temp}°C ${results.description}`,
    min: `${
      results.forecast.find((f: any) => results.date.includes(f.date)).min
    }°`,
    max: `${
      results.forecast.find((f: any) => results.date.includes(f.date)).max
    }°`,
    wind: results.wind_speedy,
    humidity: `${results.humidity}%`,
    forecast: results.forecast
      .map((cast: any) => ({
        day: cast.weekday,
        min: `${cast.min}°`,
        max: `${cast.max}°`
      }))
      .filter((_cast: any, index: number) => index <= LIMIT)
  }
}
