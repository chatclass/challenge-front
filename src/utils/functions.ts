import { forecastDay } from '../store/ducks/weather/types';

export const unique = (
  value: forecastDay,
  index: number,
  self: forecastDay[],
) => {
  return (
    self
      .map(e => new Date(e.dt_txt).getDay())
      .indexOf(new Date(value.dt_txt).getDay()) === index
  );
};

export const capitalize = (str: string) => {
  const lower = str.toLowerCase();
  return str.charAt(0).toUpperCase() + lower.slice(1);
};
