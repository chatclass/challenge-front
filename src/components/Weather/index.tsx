import React, { ReactEventHandler } from 'react'
import Heading from '../Heading'
import * as S from './styles'

type Forecast = {
  day: string
  min: string
  max: string
}

type WeatherProps = {
  city: string
  temperature: string
  min: string
  max: string
  sensation?: string
  wind: string
  humidity: string
  forecast?: Forecast[]
  handleClose?: ReactEventHandler
}

export default function Weather({
  city,
  temperature,
  min,
  max,
  sensation,
  wind,
  humidity,
  forecast,
  handleClose
}: WeatherProps) {
  return (
    <S.Wrapper>
      <Heading as="h5">{city}</Heading>

      <S.Close onClick={handleClose}>
        <svg
          xmlns="http://www.w3.org/2000/svg"
          height="24px"
          viewBox="0 0 24 24"
          width="24px"
          fill="currentColor"
        >
          <path d="M0 0h24v24H0z" fill="none" />
          <path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z" />
        </svg>
      </S.Close>

      <Heading as="h2">{temperature}</Heading>

      <S.Items>
        <S.Item>
          <svg
            xmlns="http://www.w3.org/2000/svg"
            enableBackground="new 0 0 24 24"
            height="24px"
            viewBox="0 0 24 24"
            width="24px"
            fill="currentColor"
          >
            <rect fill="none" height="24" width="24" />
            <path d="M19,15l-1.41-1.41L13,18.17V2H11v16.17l-4.59-4.59L5,15l7,7L19,15z" />
          </svg>
          <span>
            <b>{min}</b>
          </span>

          <svg
            xmlns="http://www.w3.org/2000/svg"
            enableBackground="new 0 0 24 24"
            height="24px"
            viewBox="0 0 24 24"
            width="24px"
            fill="currentColor"
          >
            <rect fill="none" height="24" width="24" />
            <path d="M5,9l1.41,1.41L11,5.83V22H13V5.83l4.59,4.59L19,9l-7-7L5,9z" />
          </svg>
          <span>
            <b>{max}</b>
          </span>
        </S.Item>
        {sensation && (
          <S.Item>
            <span>
              Sensação <b>{sensation}</b>
            </span>
          </S.Item>
        )}
      </S.Items>
      <S.Items>
        <S.Item>
          <span>
            Vento <b>{wind}</b>
          </span>
        </S.Item>
        <S.Item>
          <span>
            Himidade <b>{humidity}</b>
          </span>
        </S.Item>
      </S.Items>

      <S.Hr />

      <S.Forecast>
        {forecast?.map((item) => (
          <S.Day key={`${item.day}-${item.min}`}>
            <b>{item.day}</b>
            <span>
              {item.min} {item.max}
            </span>
          </S.Day>
        ))}
      </S.Forecast>
    </S.Wrapper>
  )
}
