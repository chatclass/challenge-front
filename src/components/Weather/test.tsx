import React from 'react'
import { render, screen } from '../../utils/test-utils'

import Weather from '.'

describe('<Weather />', () => {
  it('should render the Weather', () => {
    const { container } = render(
      <Weather
        city="são paulo, sp"
        temperature="27°C Parcialmente nublado"
        min="18°"
        max="30°"
        sensation=""
        wind="0.45 km/h"
        humidity="57%"
        forecast={[
          {
            day: 'Qui',
            min: '17°',
            max: '25°'
          }
        ]}
      />
    )

    expect(
      screen.getByRole('heading', { name: /são paulo, sp/i })
    ).toBeInTheDocument()
    expect(
      screen.getByRole('heading', {
        name: /27°c parcialmente nublado/i
      })
    ).toBeInTheDocument()
    expect(screen.getByText(/qui/i)).toBeInTheDocument()
    expect(screen.getByText(/17° 25°/i)).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })
})
