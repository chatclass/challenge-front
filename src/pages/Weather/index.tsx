import React, { useEffect, useCallback, useState, Fragment } from 'react'
import { getCapitals } from '../../api/capitals'
import Heading from '../../components/Heading'
import TextField from '../../components/TextField'
import Weather from '../../components/Weather'
import { useWeather, Weather as WeatherProps } from '../../hooks/useWeather'
import * as S from './styles'
import SearchIcon from './searchIcon'

export default function WeatherPage() {
  const { weather, setWeather, handleSearch, error } = useWeather()
  const [capitals, setCapitals] = useState<WeatherProps[]>([])

  useEffect(() => {
    const loadCapitals = async () => {
      try {
        const capitals = await getCapitals()
        setCapitals(capitals)
      } catch (e) {
        console.error(e)
      }
    }

    loadCapitals()
  }, [])

  const handleClose = useCallback(() => {
    setWeather(undefined)
  }, [setWeather])

  return (
    <S.Wrapper>
      <S.Container>
        <Heading as="h1" color="light">
          Previsão do tempo
        </Heading>
      </S.Container>

      {weather?.city && <Weather {...weather} handleClose={handleClose} />}

      <S.Container>
        <TextField
          autoFocus
          placeholder="Insira aqui o nome da cidade"
          iconPosition="right"
          icon={SearchIcon}
          onInputChange={handleSearch}
          error={error!}
        />
      </S.Container>

      {capitals.length ? (
        <>
          <S.Hr />

          <S.Container>
            <Heading as="h3" color="light">
              Capitais
            </Heading>

            <S.Dl title="Weather capitals">
              <div>
                <S.Dt>
                  <span>Min</span>
                  <span>Máx</span>
                </S.Dt>
                {capitals.map((capital: WeatherProps) => (
                  <S.Dd key={`${capital.city}-${capital.wind}`}>
                    <span>{capital.min}</span>
                    <span>{capital.max}</span>
                    <span>{capital.city}</span>
                    <span>Vento {capital.wind}</span>
                  </S.Dd>
                ))}
              </div>
            </S.Dl>
          </S.Container>
        </>
      ) : null}
    </S.Wrapper>
  )
}
