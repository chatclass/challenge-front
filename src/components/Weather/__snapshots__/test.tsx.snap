// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Weather /> should render the Weather 1`] = `
.c1 {
  font-size: 1rem;
  font-weight: 700;
  color: #000000;
}

.c3 {
  font-size: 3.5rem;
  font-weight: 700;
  color: #000000;
}

.c0 {
  padding: 1rem 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  background-color: #FCFCFDE6;
  margin-bottom: 2rem;
  box-shadow: 0px 0px 0.5rem rgba(0,0,0,0.25);
}

.c0 > h2 {
  font-size: 2.5rem;
  margin: 1.5rem 0;
  color: #353945;
}

.c0 * svg {
  color: #ff7f00;
}

.c2 {
  border: 0;
  background: transparent;
  position: absolute;
  padding: 0.25rem;
  top: 0.5rem;
  right: 0.5rem;
  cursor: pointer;
  border: 2px transparent;
}

.c2:focus-visible {
  outline-color: #ff7f00;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 > span {
  color: #777E90;
  font-size: 1rem;
}

.c5 > span > b {
  font-weight: 700;
  color: #141416;
}

.c4 {
  max-width: 16.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 2rem;
  margin-top: 1rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c6 {
  width: 100%;
  height: 1px;
  margin: 1rem 0;
  border-top: 0;
  border-bottom: 1px solid #ff7f00;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 > span {
  color: #ff7f00;
  font-size: 1rem;
}

.c8 > b {
  font-weight: 700;
  color: #141416;
}

@media (min-width:768px) {
  .c0 {
    padding: 2rem 4rem;
  }

  .c0 > h2 {
    font-size: 3.5rem;
  }
}

@media (min-width:768px) {
  .c2 {
    top: 1rem;
    right: 1rem;
  }
}

@media (min-width:768px) {
  .c6 {
    margin: 2rem 0;
  }
}

<div
  class="c0"
>
  <h5
    class="c1"
    color="dark"
  >
    são paulo, sp
  </h5>
  <button
    class="c2"
  >
    <svg
      fill="currentColor"
      height="24px"
      viewBox="0 0 24 24"
      width="24px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </button>
  <h2
    class="c3"
    color="dark"
  >
    27°C Parcialmente nublado
  </h2>
  <div
    class="c4"
  >
    <div
      class="c5"
    >
      <svg
        enable-background="new 0 0 24 24"
        fill="currentColor"
        height="24px"
        viewBox="0 0 24 24"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="24"
          width="24"
        />
        <path
          d="M19,15l-1.41-1.41L13,18.17V2H11v16.17l-4.59-4.59L5,15l7,7L19,15z"
        />
      </svg>
      <span>
        <b>
          18°
        </b>
      </span>
      <svg
        enable-background="new 0 0 24 24"
        fill="currentColor"
        height="24px"
        viewBox="0 0 24 24"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="24"
          width="24"
        />
        <path
          d="M5,9l1.41,1.41L11,5.83V22H13V5.83l4.59,4.59L19,9l-7-7L5,9z"
        />
      </svg>
      <span>
        <b>
          30°
        </b>
      </span>
    </div>
    
  </div>
  <div
    class="c4"
  >
    <div
      class="c5"
    >
      <span>
        Vento 
        <b>
          0.45 km/h
        </b>
      </span>
    </div>
    <div
      class="c5"
    >
      <span>
        Himidade 
        <b>
          57%
        </b>
      </span>
    </div>
  </div>
  <hr
    class="c6"
  />
  <div
    class="c7"
  >
    <div
      class="c8"
    >
      <b>
        Qui
      </b>
      <span>
        17°
         
        25°
      </span>
    </div>
  </div>
</div>
`;
